File: Employee.cpp
#include "Employee.hpp"
#include <iostream>

// Инициализация статических полей
int Employee::totalEmployees = 0;
double Employee::totalHoursWorked = 0.0;

Employee::Employee(int id, const std::string& name, const std::string& position, double rate)
    : employeeId(id), name(name), position(position), hoursWorked(0.0), hourlyRate(rate)
{
    try {
        if (id <= 0) {
            throw std::invalid_argument("Employee ID must be positive");
        }
        if (rate < 0.0) {
            throw std::invalid_argument("Hourly rate cannot be negative");
        }
        totalEmployees++;
    } catch (const std::exception& e) {
        std::cerr << "Error in Employee constructor: " << e.what() << std::endl;
    }
}

// Конструктор копирования
Employee::Employee(const Employee& other)
    : employeeId(other.employeeId), name(other.name), position(other.position),
      hoursWorked(other.hoursWorked), hourlyRate(other.hourlyRate)
{
    totalEmployees++;
    totalHoursWorked += other.hoursWorked;
}

Employee::~Employee() {
    // При удалении сотрудника вычитаем его часы из общего количества
    totalHoursWorked -= this->hoursWorked;
    totalEmployees--;
}

double Employee::calculateSalary() const {
    return this->hoursWorked * this->hourlyRate;
}

void Employee::addHours(double hours) {
    try {
        if (hours < 0.0) {
            throw std::invalid_argument("Hours cannot be negative");
        }
        double oldHours = this->hoursWorked;
        this->hoursWorked += hours;
        totalHoursWorked += (this->hoursWorked - oldHours);
    } catch (const std::exception& e) {
        std::cerr << "Error in addHours: " << e.what() << std::endl;
    }
}

void Employee::resetHours() {
    totalHoursWorked -= this->hoursWorked;
    this->hoursWorked = 0.0;
}

void Employee::showInfo() const {
    std::cout << "Employee Info:" << std::endl;
    std::cout << "  ID: " << this->employeeId << std::endl;
    std::cout << "  Name: " << this->name << std::endl;
    std::cout << "  Position: " << this->position << std::endl;
    std::cout << "  Hours Worked: " << this->hoursWorked << std::endl;
    std::cout << "  Hourly Rate: " << this->hourlyRate << std::endl;
    std::cout << "  Salary: " << this->calculateSalary() << std::endl;
}

// Оператор присваивания
Employee& Employee::operator=(const Employee& other) {
    if (this != &other) {
        totalHoursWorked -= this->hoursWorked; // Убираем старые часы
        this->employeeId = other.employeeId;
        this->name = other.name;
        this->position = other.position;
        this->hoursWorked = other.hoursWorked;
        this->hourlyRate = other.hourlyRate;
        totalHoursWorked += this->hoursWorked; // Добавляем новые часы
    }
    return *this;
}

// Оператор сравнения
bool Employee::operator==(const Employee& other) const {
    return this->employeeId == other.employeeId;
}

// Оператор < (сравнение по зарплате)
bool Employee::operator<(const Employee& other) const {
    return this->calculateSalary() < other.calculateSalary();
}

File: Java/Employee.java
// Абстрактный класс Employee с protected полями для наследования
public abstract class Employee implements Payable, Displayable {
    protected String name;
    protected String position;
    protected double hoursWorked;
    protected double hourlyRate;
    protected static int totalEmployees = 0;
    protected static double totalHoursWorked = 0.0;

    public Employee(String name, String position, double hourlyRate) {
        this.name = name;
        this.position = position;
        this.hoursWorked = 0.0;
        this.hourlyRate = hourlyRate;
        totalEmployees++;
    }

    // Виртуальный метод (может быть переопределен в производных классах)
    public double calculateSalary() {
        return hoursWorked * hourlyRate;
    }

    // Абстрактный метод (должен быть реализован в производных классах)
    public abstract String getEmployeeType();

    public void addHours(double hours) {
        this.hoursWorked += hours;
        System.out.println("Added " + hours + " hours. Total hours worked: " + hoursWorked);
    }

    // Виртуальный метод (может быть переопределен в производных классах)
    public void showInfo() {
        System.out.println("Employee Info:");
        System.out.println("  Name: " + name);
        System.out.println("  Position: " + position);
        System.out.println("  Hours Worked: " + hoursWorked);
        System.out.println("  Hourly Rate: " + hourlyRate);
        System.out.println("  Salary: " + calculateSalary());
    }

    // Реализация интерфейса Payable
    @Override
    public double getAmount() {
        return calculateSalary();
    }

    @Override
    public void processPayment() {
        System.out.println("Processing salary payment for employee: " + name);
    }

    @Override
    public String getPaymentInfo() {
        return "Employee " + name + " salary: " + calculateSalary();
    }

    // Реализация интерфейса Displayable
    @Override
    public void display() {
        showInfo();
    }

    @Override
    public String getDisplayString() {
        return "Employee: " + name + " - " + position;
    }

    // Статические методы
    public static int getTotalEmployees() {
        return totalEmployees;
    }

    public static double getTotalHoursWorked() {
        return totalHoursWorked;
    }

    public static double getAverageHours() {
        return totalEmployees > 0 ? totalHoursWorked / totalEmployees : 0.0;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getPosition() {
        return position;
    }

    public void setPosition(String position) {
        this.position = position;
    }

    public double getHoursWorked() {
        return hoursWorked;
    }

    public void setHoursWorked(double hoursWorked) {
        this.hoursWorked = hoursWorked;
    }

    public double getHourlyRate() {
        return hourlyRate;
    }

    public void setHourlyRate(double hourlyRate) {
        this.hourlyRate = hourlyRate;
    }
}


